Fixed #29010, Fixed #29138 -- Added limit_choices_to and to_field support to autocomplete fields

codingjoe opened the pull request:
Fixed [#29010](https://code.djangoproject.com/ticket/29010) -- Added limit_choices_to support to autocomplete fields.
Both FK and m2m fields support limit_choices_to which prev. was
ignored by autocomplete fields.

This incidentally also fixes [#29138](https://code.djangoproject.com/ticket/29138):
Fix #29138 -- Add customer to_field support to autocomplete fields.
FKs can have a custom to_field (not PK). This was not supported by
autocompelte fields.

codingjoe commented:
pull-requests-selenium

codingjoe commented:
buildbot, test on selenium

codingjoe commented:
buildbot, test on selenium.

charettes commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,12 +25,19 @@ def get(self, request, *args, **kwargs):
-        self.paginator_class = self.model_admin.paginator
+        app_label = request.GET.get('app_label')
+        model_name = request.GET.get('model_name')
+        field_name = request.GET.get('field_name')
```
The `field_name` has to be validated else a malicious user could set it to a sensitive field (e.g. `'password'` for `User.password`) and retrieve data they are not supposed to be allowed to see.

You can use `self.model_admin.to_field_allowed(request, field_name)`

e.g.

https://github.com/django/django/blob/f1c011f9592ef5dc99a4a3d7c54ad30fe0d722c5/django/contrib/admin/options.py#L1822-L1824

A test should be added to make sure this is appropriately validated.

codingjoe replied:
I added the check and a test to verify it can not be misused.

charettes commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,12 +25,19 @@ def get(self, request, *args, **kwargs):
         self.term = request.GET.get('term', '')
-        self.paginator_class = self.model_admin.paginator
+        app_label = request.GET.get('app_label')
+        model_name = request.GET.get('model_name')
```
Are these really necessary? Can't they be retrieved from `self.object_list.model`?

codingjoe replied:
No, this is not. they are different models. Once being the source with the FK, the other the model the FK points to.

charettes replied:
I see. Any thoughts about passing the _local_ field instead of the remote one and the model it's from?

e.g. in the case of tests below `field_name` would be `'uuid'` instead of `'question_with_to_field'`.

That's how the other `to_field` enabled admin views work and what `to_field_allowed` expects. It also feels like a nicer API to me and would allow you to stop passing `app_label` and `model_name`.

codingjoe replied:
@charettes the `field_name` would be `question_with_to_field` os the local field. See the new tests cases I added.

codingjoe replied:
In any case, naming could be improved I guess to avoid the confusion.

charettes replied:
@codingjoe it's the remote field in test and not the local one.

The URL is `'autocomplete_admin:admin_views_question_autocomplete'` (which is bound to the `Question` model) and `'question_with_to_field'` is an `Answer` field.

What I'm suggesting we do instead is pass `field='uuid'` which is a field local to `Question` and have `Answer` admin pass `field.remote_field.name` instead which is `'uuid'`.

That's what the other admin views are doing and doesn't require you do extra `model` and `app_label` sanitization when walking your way back to the local field from the remote one.

apollo13 replied:
I agree with Simon here, if you for instance look at the edit popup next to foreignkeys you will see this in the URL: `http://127.0.0.1:8000/admin/auth/user/apo/change/?_to_field=username&_popup=1`

So I propose that the autocomplete request should look like this: `http://127.0.0.1:8000/admin/auth/user/autocomplete/?_to_field=username`

and not like currently: `http://127.0.0.1:8000/admin/auth/user/autocomplete/?app_label=select2&model_name=test&field_name=user`

codingjoe replied:
As much as I want to agree with the two of you, I can't. The solution you are proposing only solves the `to_field` issue #29138 but not #29010 which is about `limit_choices_to`. For that to work I need to local field and not only the `to_field`.

To be honest here, the `to_field` issues, is somewhat of an edge case, I see the `limit_choices_to` problem as much more severe, since it's a more commonly used feature.

charettes commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,12 +25,19 @@ def get(self, request, *args, **kwargs):
             return JsonResponse({'error': '403 Forbidden'}, status=403)
 
         self.term = request.GET.get('term', '')
-        self.paginator_class = self.model_admin.paginator
```
This should be removed in a separate commit if it's unused.

codingjoe replied:
haha, can't sneak anything past you. Very good point ;)

codingjoe commented:
buildbot, test on selenium.

codingjoe commented:
buildbot, test on selenium.

codingjoe commented:
@charettes do you think now that 2.2 is out we could get this on the road. It's kind of sad that it didn't make it into 2.2 because this is a bug fix or improved implementation of an existing feature, but better now than never.

apollo13 commented:
Hrmpf, github does not let you unresolve discussion, either way here is my change request: https://github.com/django/django/pull/11026/files/5a93defb53cac18d6a0e1c28e07dd8bea79c3e67#r275113897

apollo13 commented:
Also if I try this patch with the following admin & model:
```
from django.db import models
from django.contrib.auth.models import User


class Test(models.Model):
    title = models.TextField(blank=True)
    user = models.ForeignKey(User, to_field="username", on_delete=models.CASCADE)
```

```
from django.contrib import admin

from .models import Test


class TestAdmin(admin.ModelAdmin):
    autocomplete_fields = ["user"]


admin.site.register(Test, TestAdmin)
```

and then create and object and open it in the admin I get:

```
Environment:


Request Method: GET
Request URL: http://localhost:8000/admin/select2/test/2/change/

Django Version: 3.0.dev20190309115740
Python Version: 3.7.3
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'select2']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']


Template error:
In template /home/florian/sources/django.git/django/contrib/admin/templates/admin/includes/fieldset.html, error at line 19
   invalid literal for int() with base 10: 'apo'
   9 :             {% for field in line %}
   10 :                 <div{% if not line.fields|length_is:'1' %} class="fieldBox{% if field.field.name %} field-{{ field.field.name }}{% endif %}{% if not field.is_readonly and field.errors %} errors{% endif %}{% if field.field.is_hidden %} hidden{% endif %}"{% elif field.is_checkbox %} class="checkbox-row"{% endif %}>
   11 :                     {% if not line.fields|length_is:'1' and not field.is_readonly %}{{ field.errors }}{% endif %}
   12 :                     {% if field.is_checkbox %}
   13 :                         {{ field.field }}{{ field.label_tag }}
   14 :                     {% else %}
   15 :                         {{ field.label_tag }}
   16 :                         {% if field.is_readonly %}
   17 :                             <div class="readonly">{{ field.contents }}</div>
   18 :                         {% else %}
   19 :                              {{ field.field }} 
   20 :                         {% endif %}
   21 :                     {% endif %}
   22 :                     {% if field.field.help_text %}
   23 :                         <div class="help">{{ field.field.help_text|safe }}</div>
   24 :                     {% endif %}
   25 :                 </div>
   26 :             {% endfor %}
   27 :         </div>
   28 :     {% endfor %}
   29 : </fieldset>


Traceback:

File "/home/florian/sources/django.git/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/florian/sources/django.git/django/core/handlers/base.py" in _get_response
  145.                 response = self.process_exception_by_middleware(e, request)

File "/home/florian/sources/django.git/django/core/handlers/base.py" in _get_response
  143.                 response = response.render()

File "/home/florian/sources/django.git/django/template/response.py" in render
  106.             self.content = self.rendered_content

File "/home/florian/sources/django.git/django/template/response.py" in rendered_content
  83.         content = template.render(context, self._request)

File "/home/florian/sources/django.git/django/template/backends/django.py" in render
  61.             return self.template.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  171.                     return self._render(context)

File "/home/florian/sources/django.git/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/florian/sources/django.git/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/florian/sources/django.git/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/defaulttags.py" in render
  209.                     nodelist.append(node.render_annotated(context))

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/loader_tags.py" in render
  188.             return template.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  173.                 return self._render(context)

File "/home/florian/sources/django.git/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/defaulttags.py" in render
  209.                     nodelist.append(node.render_annotated(context))

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/defaulttags.py" in render
  209.                     nodelist.append(node.render_annotated(context))

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/defaulttags.py" in render
  309.                 return nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/defaulttags.py" in render
  309.                 return nodelist.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/florian/sources/django.git/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/florian/sources/django.git/django/template/base.py" in render
  993.         return render_value_in_context(output, context)

File "/home/florian/sources/django.git/django/template/base.py" in render_value_in_context
  972.             value = str(value)

File "/home/florian/sources/django.git/django/utils/html.py" in <lambda>
  388.     klass.__str__ = lambda self: mark_safe(klass_str(self))

File "/home/florian/sources/django.git/django/forms/boundfield.py" in __str__
  33.         return self.as_widget()

File "/home/florian/sources/django.git/django/forms/boundfield.py" in as_widget
  93.             renderer=self.form.renderer,

File "/home/florian/sources/django.git/django/forms/widgets.py" in render
  241.         context = self.get_context(name, value, attrs)

File "/home/florian/sources/django.git/django/contrib/admin/widgets.py" in get_context
  287.             'rendered_widget': self.widget.render(name, value, attrs),

File "/home/florian/sources/django.git/django/forms/widgets.py" in render
  241.         context = self.get_context(name, value, attrs)

File "/home/florian/sources/django.git/django/forms/widgets.py" in get_context
  680.         context = super().get_context(name, value, attrs)

File "/home/florian/sources/django.git/django/forms/widgets.py" in get_context
  641.         context['widget']['optgroups'] = self.optgroups(name, context['widget']['value'], attrs)

File "/home/florian/sources/django.git/django/contrib/admin/widgets.py" in optgroups
  431.             for obj in self.choices.queryset.using(self.db).filter(pk__in=selected_choices)

File "/home/florian/sources/django.git/django/db/models/query.py" in filter
  881.         return self._filter_or_exclude(False, *args, **kwargs)

File "/home/florian/sources/django.git/django/db/models/query.py" in _filter_or_exclude
  899.             clone.query.add_q(Q(*args, **kwargs))

File "/home/florian/sources/django.git/django/db/models/sql/query.py" in add_q
  1272.         clause, _ = self._add_q(q_object, self.used_aliases)

File "/home/florian/sources/django.git/django/db/models/sql/query.py" in _add_q
  1300.                     split_subq=split_subq, simple_col=simple_col,

File "/home/florian/sources/django.git/django/db/models/sql/query.py" in build_filter
  1233.         condition = self.build_lookup(lookups, col, value)

File "/home/florian/sources/django.git/django/db/models/sql/query.py" in build_lookup
  1098.         lookup = lookup_class(lhs, rhs)

File "/home/florian/sources/django.git/django/db/models/lookups.py" in __init__
  20.         self.rhs = self.get_prep_lookup()

File "/home/florian/sources/django.git/django/db/models/lookups.py" in get_prep_lookup
  210.                 rhs_value = self.lhs.output_field.get_prep_value(rhs_value)

File "/home/florian/sources/django.git/django/db/models/fields/__init__.py" in get_prep_value
  968.         return int(value)

Exception Type: ValueError at /admin/select2/test/2/change/
Exception Value: invalid literal for int() with base 10: 'apo'
```

codingjoe commented:
> 



> Hrmpf, github does not let you unresolve discussion, either way here is my change request: https://github.com/django/django/pull/11026/files/5a93defb53cac18d6a0e1c28e07dd8bea79c3e67#r275113897

I unresolved it, but it still isn't a viable solution :P

Thanks for the crash report. I did find the error, I am working on it, right now.

apollo13 commented:
@codingjoe You fixed the crash (I guess) but didn't add any tests, which means the code now works by "accident" and still isn't tested. I'll look over the limit_choices_to vs to_field and think about it.

apollo13 commented:
And the fact that you put two changes in one PR doesn't make reviewing any easier :D

apollo13 commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,11 +26,29 @@ def get(self, request, *args, **kwargs):
             return JsonResponse({'error': '403 Forbidden'}, status=403)
 
         self.term = request.GET.get('term', '')
+        app_label = request.GET.get('app_label')
```
This line and the following ones will throw a server error if the variables do not exist?

apollo13 commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,11 +26,29 @@ def get(self, request, *args, **kwargs):
+        model_name = request.GET.get('model_name')
+        field_name = request.GET.get('field_name')
+
+        model = apps.get_model(app_label, model_name)
```
This can raise a `LookupError`

apollo13 commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,11 +26,29 @@ def get(self, request, *args, **kwargs):
+        model = apps.get_model(app_label, model_name)
+
+        try:
+            self.field = model._meta.get_field(field_name)
```
If I understood the above code correctly, then `self.field` is on the source model, whereas `self.model_admin` points to the target admin, I think we should really rename those to avoid confusion.

apollo13 commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -24,11 +26,29 @@ def get(self, request, *args, **kwargs):
+        if self.field.remote_field is None or self.field.remote_field.model != self.model_admin.model:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        to_field_name = getattr(self.field.remote_field, 'field_name', self.model_admin.model._meta.pk.name)
```
From reading through Django's source code, you can rely that `self.field_remote_field.field_name` is set I think:

https://github.com/django/django/blob/a8b3f96f6acfa082f99166e0a1cfb4b0fbc0eace/django/db/models/fields/related.py#L945-L948

codingjoe replied:
This code is executed only for `ForeignKey`s not M2M or any other custom `RelatedObject` a developer might create.

apollo13 commented:
What about m2m fields?

apollo13 commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -40,7 +60,12 @@ def get_paginator(self, *args, **kwargs):
 
     def get_queryset(self):
         """Return queryset based on ModelAdmin.get_search_results()."""
+        limit_choices_to = self.field.remote_field.limit_choices_to
```
Isn't this and the following lines essentially the same as `limiti_choices_to = self.field.get_limit_choices_to()`?

codingjoe commented:
buildbot, test on selenium.

codingjoe commented:
Hey @apollo13 it's been a while, but I updated the patch. I changed a test to catch your crash and I renamed some things to make the code more readable. The biggest change is, that I don't need a view per model admin anymore. Per site is just fine, since we pass the model admin information via get parameter anyways.

> And the fact that you put two changes in one PR doesn't make reviewing any easier :D

One just happened to solve the other. The focus is on `limit_choices_to`. It requires access to the actual model field, that `to_field` does not.

codingjoe commented:
buildbot, test on selenium.

karbassi commented:
@apollo13 @charettes @codingjoe It would be amazing to get this reviewed and merged. Thanks for all you all do! ❤️

codingjoe commented:
Thanks @karbassi for your kind words. Yes, it would be amazing, but we are always short on reviews. Maybe you give it a spin. You don't need to be a core developer to review someone's code. This would certainly help move things along.

I just rebased my branch to the latest master.

xi commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -15,6 +17,34 @@ def get(self, request, *args, **kwargs):
+        try:
+            self.model_admin = self.admin_site._registry[remote_model]
+        except KeyError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
```
Why 403 instead of 404? If there is a good reason, maybe a commend would be helpful.

codingjoe replied:
How so, all errors lead to a 403-response, not a 404. Therefore, this seems consistent.

xi replied:
Sorry if my comment was misleading. My point is that **all** of these should be 404 because 403 is mostly for permission checks.

If you look at the old code there were two checks: One about `get_search_fields` and one about permissions. The former used 404 and the latter 403.

xi commented:
> > And the fact that you put two changes in one PR doesn't make reviewing any easier :D
>
> One just happened to solve the other. The focus is on `limit_choices_to`. It requires access to the actual model field, that `to_field` does not.

As far as I understand this could be split into 3 commits:

- Make model field available in `AutocompleteJsonView` without changing behavior
- Fix `to_field`
- Fix `limit_choices_to`

For me it is hard to understand the changes to the tests. Maybe splitting it like this makes it easier to understand those test changes.

codingjoe commented:
@xi thanks your review, much appreciated. I am not so sure splitting this into separate commits, is a good idea. I agree that this change fixes multiple issues, but it is only a single change (even if a larger one). Splitting everything is not only work, but might make it harder for others to revert changes or refactor things in the future. Nevertheless, I'd be glad to do it, but I would leave the decision to @felixxm 

@xi you are otherwise happy, feel free to move triage to ready for check-in, if you care about this being released soon.

carltongibson reviewed (commented):
Hey @codingjoe — Sorry for the slow uptake here: it's been one of those scary lingerers. Either way, you've finally made it to the top (or bottom, depending on how you look at it) of my list. 🙂

Initially it looks good. I think we will need some release notes, just to say, _now supports_ and point out the change in the URL. (Nobody should be using that but...) 

I need to sit down and think through each of the test cases but am commenting now in case you have capacity to give it a look over and think if there are any extra tests or docs you'd like to add. 

Thanks!

codingjoe commented:
@carltongibson thanks mate, I really appreciate your help here. I have too many scary open PRs that are lost in review abyss.
Feel free to make changes directly. Release notes, tests, whatever makes this move along more quickly.

ababic commented on django/contrib/admin/widgets.py:
```diff
@@ -380,18 +380,18 @@ class AutocompleteMixin:
 
-    def __init__(self, rel, admin_site, attrs=None, choices=(), using=None):
-        self.rel = rel
+    def __init__(self, field, admin_site, attrs=None, choices=(), using=None):
```
Hi @codingjoe. `self.choices` should be a` ModelChoiceIterator` instance, so `self.field` can be derived from `self.choices.field`. With this in mind, is the new `field` init argument really necessary?

ababic replied:
Not sure who marked this as resolved, because it isn't.

carltongibson replied:
AdminSelect used to take the related model. It now takes the field.

kakulukia commented:
What about also supporting `ModelMultipleChoiceField.label_from_instance`?

ababic commented:
Hi @kakulukia.

> What about also supporting ModelMultipleChoiceField.label_from_instance?

That would be quite difficult, as the view generating the choices is being called via ajax, and although it can look up the relevant model field from the model/field names in the URL, there's no reliable way to access the `ModelChoiceIterator`, as that is generated by a 'form field', and there's currently no way to know which form field the choices are being generated for.

This wouldn't be impossible, but would add quite a bit of complexity (it would require of 'registry' of widget instances,  that the autocomplete view could use to lookup with widget instance via an ID passed in the URL instead)

kakulukia commented:
Yes, i saw this yesterday as i tried to solve that myself. While the rest of this patch works like a charm, having to get to the model form field is much more difficult than to the actual DB field.
Plus changing the form field also removes the autocompletion widget and figuring out  haw to bring it back, took me some time. :(

@ababic the choices are empty - probably because they are loaded by JS, hence the extra field argument.

carltongibson commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -15,6 +17,34 @@ def get(self, request, *args, **kwargs):
         """
+        self.term = request.GET.get('term', '')
+
+        try:
+            app_label = request.GET['app_label']
+            model_name = request.GET['model_name']
+            field_name = request.GET['field_name']
+        except KeyError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            source_model = apps.get_model(app_label, model_name)
+        except LookupError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            self.source_field = source_model._meta.get_field(field_name)
+        except FieldDoesNotExist:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            remote_model = self.source_field.remote_field.model
+        except AttributeError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+        try:
+            self.model_admin = self.admin_site._registry[remote_model]
+        except KeyError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
```
Hey @codingjoe — so I think I would restructure this section slightly, through to line 59. 

* I'd create a method to configure three attributes on self: `source_field`, `model_admin`, and `to_field_name`. It'd need to take the `request`. In there I'd include all the `try:except:` blocks and raise an `AutoCompleteSetupError` if any of them occur. 
* Outside that method I would have a single `try:except:` to log the error, and return the JSON 403 response. (I'm thinking about how I'd debug this if there's no logging, and all I have to go on is a 403...) 

I think we should also move the `has_perm() -> 403` check **above** the `get_search_fields() -> 404`. (That last if nicely informative for users but we shouldn't hand out that info to someone without the permissions to access the modeladmin at all… 🤔)

codingjoe replied:
Very good idea, I moved the code into a separate method. I further agree with the problem of debugging potential errors. When I originally submitted the patch we switch to a 403 JSON response, to have a consistent response content type. However, that is not necessary, since Select2 does not care about the content type. If the response code is 4xx or 5xx it will always display the same message and not even try to decode the content. Therefore, I believe it's best to simply raise a `PermissionDenied` and let the middleware handle logging. I added proper `raise from` clauses, to provide users with additional information.

I kept the `has_perm` call in the `get` method, to make sure it doesn't get lost and I don't find it logically connected to the rest.

carltongibson commented on django/contrib/admin/widgets.py:
```diff
@@ -380,18 +380,18 @@ class AutocompleteMixin:
-        self.rel = rel
+    def __init__(self, field, admin_site, attrs=None, choices=(), using=None):
+        self.field = field
+        self.rel = field.remote_field
```
`self.rel` looks unused now. Can we remove it?

carltongibson reviewed (commented):
Hey @codingjoe — I have a suggestion for a refactoring of the view set-up code. What do you think about that? Do you have capacity to look at it? 

Other than that (and docs we mentioned, which I can help with) it looks sensible. 
Good stuff.

felixxm commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -41,6 +74,7 @@ def get_paginator(self, *args, **kwargs):
     def get_queryset(self):
         """Return queryset based on ModelAdmin.get_search_results()."""
         qs = self.model_admin.get_queryset(self.request)
+        qs = qs.complex_filter(self.source_field.get_limit_choices_to())
```
As far as I'm aware, this can create duplicates on multi-value relations, see related fix for forms 556fa4bbba5ba86bc1646a86fb11ab55405d4aa4.

codingjoe replied:
Yes, it does. This is handled two lines below.

felixxm replied:
It looks like this can still produce duplicates, see [comment](https://code.djangoproject.com/ticket/15819#comment:16). But that's a separate issue.

codingjoe replied:
Hm… I see, but yes, that is a separate issue.

felixxm commented on tests/admin_views/test_autocomplete_view.py:
```diff
@@ -63,19 +82,70 @@ def test_success(self):
+
+    def test_limit_choices_to(self):
+        q = Question.objects.create(question='Is this a question?')
+        _ = Question.objects.create(question='Not a question.')
```
```suggestion
        Question.objects.create(question='Not a question.')
```

carltongibson commented on tests/admin_views/models.py:
```diff
@@ -629,13 +629,21 @@ class Question(models.Model):
+        related_name='uuid_answers',
+        limit_choices_to=~models.Q(question__istartswith='not'),
+    )
+    related_answers = models.ManyToManyField('self')
```
Do we have a test example for this field?

codingjoe replied:
It is being used for a negative test. The field exists, but is not included in the `autocomplete_fields`, thus the view should not work for that field, and it doesn't.

codingjoe commented:
buildbot, test on selenium.

codingjoe commented:
@carltongibson I kept the changes in a separate commit, as always, feel free squash and amend any changes you deem necessary.
BTW, I am a big clueless on how to improve the docs, since the patch does not include any public API changes or additions. Things just work as they were supposed to from the beginning 🤷

kakulukia commented:
Thx @codingjoe for fixing the final issues. Hope this makes its way into 3.2. :)

codingjoe commented:
> Thx @codingjoe for fixing the final issues. Hope this makes its way into 3.2. :)

So do I ;)

carltongibson reviewed (approved):
Hey @codingjoe. Thanks for this, really good. 

I squashed and rebased, and added a release note. (On that last, I didn't mention the move of the view to the AdminSite, since that's an implementation detail I think.)

🏅

felixxm commented:
buildbot, test on selenium.

felixxm commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
+        try:
+            model_admin = self.admin_site._registry[remote_model]
+        except KeyError as e:
+            raise PermissionDenied from e
```
I think we should raise `BadRequest` instead of `PermissionDenied`, because missing parameter in `GET` has nothing to do with permissions. Maybe, I'm picky :shrug:

carltongibson replied:
The issue here is that we can't do the proper permission check (`if not self.has_perm(request) ...`) until after we've assigned `self.model_admin`, which we return from this method. 

_Maybe_ it's OK. But… it seems like we have the exact kind of _"Folks can probe the autocomplete view to find out the structure of the application"_ issue that we spend a lot of time thinking about and trying to avoid. 

If we just raise `PermissionDenied` for all of these cases we don't leak any info. Such exceptions are logged and a developer can use that to debug. Given that this is an ajax view, they're not going to see the debug output page anyway (without taking steps). 

I'm not convinced the different status code is worth the additional exposure (or even the effort justifying why it's OK). 

> 🤷 

Yes. 🙂

xi replied:
This sounds like a non-obvious justification. Maybe it would be good to add a comment for future developers?

felixxm commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
+        if not model_admin.get_search_fields(request):
+            raise Http404(
+                '%s must have search_fields for the autocomplete_view.' %
+                type(model_admin).__name__
```
```suggestion
                type(model_admin).__qualname__
```

felixxm commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
+            Http404: If the target model admin is not configured properly.
+
+        """
+        # extract request parameters
```
IMO we can chop this comment.

felixxm commented on django/contrib/admin/views/autocomplete.py:
```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
+        except KeyError as e:
+            raise PermissionDenied from e
+
+        # retrieve Django objects from parameters
```
"Django" doesn't add much:
```suggestion
        # Retrieve objects from parameters.
```

felixxm commented on tests/admin_views/test_autocomplete_view.py:
```diff
@@ -53,7 +72,8 @@ def setUpTestData(cls):
     def test_success(self):
         q = Question.objects.create(question='Is this a question?')
-        request = self.factory.get(self.url, {'term': 'is'})
+
```
Chop blank line.

carltongibson commented:
Hey @felixxm — I resolved the nitpicks, and replied inline about the 400/403 question. Not sure what I think on the last.

dnk8n commented:
When will this be released? Doesn't seem to be working currently as I still get the error this is meant to fix (using Django 3.1.6)

carltongibson commented:
It is part of Django 3.2, which is due for final release in April.
